[dlfcn-win32] [getopt-win32] [pthreads] Improvements in package configurations

tdcosta100 opened the pull request:
**Describe the pull request**

- #### What does your PR fix?

  - **dlfcn-win32**
    For triplets with this command:

    set(VCPKG_BUILD_TYPE release)

    Installing dlfcn-win32 will fail. This PR fixes it.

  - **getopt-win32**
    This port cannot be installed as a package. This PR brings this functionality, making easier to use it.

  - **pthreads**
    Added a patch to fix EACCESSVIOLATION in pthread_getname_np.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
No

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

tdcosta100 committed:
Fix dlfcn-win32 install failure on release-only buids

tdcosta100 committed:
Update dlfcn-win32 port version

tdcosta100 commented:
@microsoft-github-policy-service agree

tdcosta100 committed:
[getopt-win32] Make package configuration

tdcosta100 committed:
[getopt-win32] Update port version

tdcosta100 committed:
[pthreads] Patch for pthread_getname_np null checking

tdcosta100 committed:
[pthreads] Update package version

tdcosta100 committed:
Fix patch encoding and line ending

tdcosta100 committed:
[getopt-win32] Fix line endings

tdcosta100 committed:
[getopt-win32] Fix code

tdcosta100 committed:
[getopt-win32] Fix typo

MonicaLiu0311 commented on ports/dlfcn-win32/portfile.cmake:
```diff
@@ -14,12 +14,14 @@ vcpkg_cmake_configure(
-file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
-file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
+if(NOT VCPKG_BUILD_TYPE STREQUAL "release")
+    file(READ "${CURRENT_PACKAGES_DIR}/debug/share/dlfcn-win32/dlfcn-win32-targets-debug.cmake" dlfcn-win32_DEBUG_MODULE)
+    string(REPLACE "\${_IMPORT_PREFIX}" "\${_IMPORT_PREFIX}/debug" dlfcn-win32_DEBUG_MODULE "${dlfcn-win32_DEBUG_MODULE}")
+    file(WRITE "${CURRENT_PACKAGES_DIR}/share/dlfcn-win32/dlfcn-win32-targets-debug.cmake" "${dlfcn-win32_DEBUG_MODULE}")
```
Recommend this function: [vcpkg_replace_string()](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_replace_string.md)

tdcosta100 replied:
Done, thank you.

tdcosta100 committed:
[dlfcn-win32] Use vcpkg_replace_string

tdcosta100 committed:
[dlfcn-win32] Install debug targets

tdcosta100 committed:
[dlfcn-win32] Fix typo

dg0yt commented on ports/getopt-win32/usage:
```diff
@@ -0,0 +1,4 @@
+The package getopt-win32 is compatible with built-in CMake targets:
+
+    find_package(getopt-win32 REQUIRED)
+    target_link_libraries(main PRIVATE getopt)
```
I see the intention. But it is unofficial. Cmake package name and target name should be prefixed accordingly.
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#add-cmake-exports-in-an-unofficial--namespace.
I also wonder about the best balance between hiding cross-platform differences ("getopt") or emphasising them ("getopt-win32").

tdcosta100 replied:
Hmmm right. So maybe I should leave only as getopt-win32, and not getopt? That's ok for me, I'm just reproducing the same as pthreads, but maybe pthreads is not that right too? Just to get acknowledged with the standards.

dg0yt replied:
Port `pthreads` is a difficult example, due to the port history in vcpkg (requiring to retain compatibility with inferior patterns from the past). In addition, there is a native threading API in windows, and there is an official `FindThreads.cmake` module which supports both pthread compatible and win32 libraries.

IMO if you want a CMake package, you should simply export `unofficial-getopt-win32-getopt.cmake` with target `unofficial::getopt-win32::getopt`, and document usage like this:
~~~
getopt-win32 provides CMake targets:

    if(WIN32 AND NOT MINGW)
        find_package(getopt-win32 REQUIRED)
        target_link_libraries(main PRIVATE getopt)
    endif()
~~~
i.e. the CMake target(s) are not "built-in" (into CMake), and it is required only for one platform.

tdcosta100 replied:
This approach is ok to my use of this package. And good to know `pthreads` is not a good example, I will make the changes for achieving this result in this package.

dg0yt commented on ports/getopt-win32/vcpkg-cmake-wrapper-getopt.cmake:
```diff
@@ -0,0 +1,10 @@
+_find_package(getopt-win32)
+set(getopt_INCLUDE_DIR "${getopt-win32_INCLUDE_DIR}")
+set(getopt_LIBRARY "${getopt-win32_LIBRARY}")
+set(getopt_LIBRARIES "${getopt-win32_LIBRARY}")
+
+if(getopt-win32_FOUND)
+    set(getopt_FOUND TRUE)
+
+    add_library(getopt ALIAS getopt-win32)
+endif()
```
If you control the package config files, there is no need for a wrapper.
Wrappers are mostly useful for official CMake modules.

tdcosta100 replied:
Yeah, makes sense if I'm not going to have a find_package(getopt REQUIRED), which is this file is intended to (be copied to share/getopt/vcpkg-cmake-wrapper.cmake). I can get rid of it if it's not right to be this way.

tdcosta100 replied:
But I have another question: if I don't create a package, I have some trouble finding the correct library (debug/release), because both have the same name. Making a package solves this just adding the package (it correctly adds debug/optimized versions to the target, I don't need to bother with it).

tdcosta100 replied:
Perfect, I will make the appropriate changes.

dg0yt commented on ports/dlfcn-win32/portfile.cmake:
```diff
@@ -14,12 +14,13 @@ vcpkg_cmake_configure(
-file(READ "${CURRENT_PACKAGES_DIR}/debug/share/dlfcn-win32/dlfcn-win32-targets-debug.cmake" dlfcn-win32_DEBUG_MODULE)
-string(REPLACE "\${_IMPORT_PREFIX}" "\${_IMPORT_PREFIX}/debug" dlfcn-win32_DEBUG_MODULE "${dlfcn-win32_DEBUG_MODULE}")
-file(WRITE "${CURRENT_PACKAGES_DIR}/share/dlfcn-win32/dlfcn-win32-targets-debug.cmake" "${dlfcn-win32_DEBUG_MODULE}")
+if(NOT VCPKG_BUILD_TYPE STREQUAL "release")
```
It is enough to write `if(NOT VCPKG_BUILD_TYPE)`

tdcosta100 replied:
But if VCKPG_BUILD_TYPE is set to release, this portion of code will fail if I leave just `if(NOT VCPKG_BUILD_TYPE)`. It removes the debug files, and in the previous version it was failing exactly because it was not enclosed by this if.

dg0yt replied:
In portfiles, you can assume that `VCKPG_BUILD_TYPE` is either false (unset, empty; implying debug+release) or `release`. Nothing else works on a relevant scale. So the proposed `if` just does the right thing, and it is generally used for debug variant blocks in portfiles.

dg0yt reviewed (commented):
I think it would be better to split the PR for each port.

Port getopt-win32 uses CMake. If you want a CMake package, simply export config from the build instead of manually authoring a config file.

tdcosta100 committed:
[pthreads] Better install handling

tdcosta100 committed:
[pthreads] Update version with corrections

tdcosta100 committed:
[dlfcn-win32] Add port-version to vcpkg.json

tdcosta100 commented:
> I think it would be better to split the PR for each port.
> 

You are right, I will try to submit different branches of my fork, so I can make a specific PR to each package.

tdcosta100 commented:
@MonicaLiu0311 

> `ports/getopt-win32/CMakeLists.txt` and `ports/getopt-win32/portfile.cmake` seems to have the line breaks changed.

I discovered some files I changed were with other line endings than most files (LF), so I fixed them, but accidentally changed line endings for files which in my system were CR LF. I will try to keep them consistent.
